src/shared/common/vector_functions.h

```diff
@@ -105,6 +105,32 @@ template<> inline DeviceVec3d VecdZero() { return { static_cast<DeviceReal>(0.0)
 template<> inline Vec2d VecdZero() { return Vec2d::Zero(); }
 template<> inline Vec3d VecdZero() { return Vec3d::Zero(); }
 
+/* specialization for specific vector operations */
+template<class RealType, int Dimension>
+inline RealType VecdDot(const sycl::vec<RealType,Dimension>& v1, const sycl::vec<RealType,Dimension>& v2) {
+    return sycl::dot(v1, v2);
+}
+template<class RealType, int Dimension>
+inline RealType VecdDot(const Eigen::Matrix<RealType,Dimension,1>& v1, const Eigen::Matrix<RealType,Dimension,1>& v2) {
+    return v1.dot(v2);
+}
+template<class RealType, int Dimension>
+inline RealType VecdNorm(const sycl::vec<RealType,Dimension>& vec) {
+    return sycl::length(vec);
+}
+template<class RealType, int Dimension>
+inline RealType VecdNorm(const Eigen::Matrix<RealType,Dimension,1>& vec) {
+    return vec.norm();
+}
+template<class RealType, int Dimension>
+inline RealType VecdSquareNorm(const sycl::vec<RealType,Dimension>& vec) {
+    return sycl::dot(vec, vec);
+}
+template<class RealType, int Dimension>
+inline RealType VecdSquareNorm(const Eigen::Matrix<RealType,Dimension,1>& vec) {
+    return vec.squaredNorm();
+}
+
```

Comment:
I am curious that which operator is faster? for example sycl or eigen?

Reply:
As I was saying, a direct comparison is not possible since Eigen is not executable within SYCL kernels, at most it would be possible to compare Eigen running on CPU vs SYCL running either on GPU or CPU, but it would not be a fair comparison.

All I can say is that [SYCL geometric functions](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:geometric-functions) will map to GPU built-in intrinsic operations, which are supposedly faster than Eigen's. But Eigen offers more operations than SYCL. For instance, Eigen has already an optimized method for transposing a matrix while for SYCL you would need to implement one yourself, which would probably be slower compared to Eigen. Nevertheless, GPU processing power will probably still outweigh Eigen's optimizations on CPU.

Reply:
Did you notice info in this link? https://developer.codeplay.com/products/computecpp/ce/2.10.0/guides/eigen-guide/options-for-building-eigen 
Seems that eigen has a SYCL back-end.

Reply:
I did, but it seems that it's limited to tensor operations and, as far as I can tell, SYCL is used as a back-end for accelerating the execution of certain methods, not to support Eigen execution within SYCL kernels.

I am now having a specific look at what they upstreamed, but from a quick glance it seems that they re-implemented certain methods to be executed by SYCL, using raw pointers to allocated device memory. A similar approach to what I was proposing. Still, there might be some of those methods that may be reused.

Reply:
Thx for the info.